### PR TITLE
Cache and reuse archive uploads

### PR DESCRIPTION
**Pull request by leeweisberger**
If two tasks use the same root (e.g. same package.json), we only want to archive and upload once and use the same upload for both builds.

---

I imagine this will be very common, so just adding in the optimization now.

### CODE REVIEW

**Review by joshma — APPROVED** [review]

**Comment by joshma** [comment]
cc @colinking on this - sanity check this is OK? this means we'd use the same upload across 2 builds, which seems fine?

File: pkg/build/remote.go [hunk 1]
```diff
@@ -185,7 +208,27 @@ func archiveTaskDir(def definitions.Definition, root string, archivePath string)
 	return nil
 }
 
-func uploadArchive(ctx context.Context, client *api.Client, taskSlug, archivePath string, loader *logger.Loader) (string, error) {
+func (d *Deployer) uploadArchive(ctx context.Context, client *api.Client, taskSlug, archivePath, rootPath string, loader *logger.Loader) (string, error) {
```
Comment (colinking) [inline pkg/build/remote.go:211]: Would recommend using a `singleflight` here: https://pkg.go.dev/golang.org/x/sync/singleflight

**Review by colinking — APPROVED** [review]
Yep, I don't see any issues with re-using. Good call @leeweisberger!

File: pkg/build/remote.go (hunk 1 above)
Comment (leeweisberger) [reply pkg/build/remote.go:205]: Yes! That was the coolest package I've ever used it fit this perfectly

Comment (leeweisberger) [reply pkg/build/remote.go:205]: I thought I had such a good idea :sob:

**Pull request merged**